Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs

```diff
@@ -239,5 +194,69 @@ private IEnumerable<Type> GetChildTypes(Type type)
 
             return allChildTypes[type];
         }
+
+        private IEnumerable<CandidateChildType> GetCandidateChildTypes(Type objectType, HashSet<string> jsonProperties)
+        {
+            var childTypes = GetChildTypes(objectType);
+            var candidateChildTypes = new List<CandidateChildType>();
+            foreach (var childType in childTypes)
+            {
+                var childTypeProperties = childType.GetProperties().Select(t => t.Name).ToList();
+                var childTypeFields = childType.GetFields().Select(t => t.Name).ToList();
+                var childTypePropertiesAndFields = new HashSet<string>(
+                    childTypeProperties.Concat(childTypeFields),
+                    StringComparer.OrdinalIgnoreCase);
+                if (jsonProperties.All(p => childTypePropertiesAndFields.Contains(p)))
+                {
+                    var candidateChildType = new CandidateChildType
+                    {
+                        Type = childType,
+                        PropertiesAndFields = childTypePropertiesAndFields
+                    };
+                    candidateChildTypes.Add(candidateChildType);
+                }
+            }
+
+            return candidateChildTypes;
+        }
+
+        private void DeserializeCandidates(JsonSerializer serializer, JObject jsonObject, IEnumerable<CandidateChildType> candidateChildTypes)
+        {
+            candidateChildTypes = candidateChildTypes.ToList();
+            foreach (var candidateChildType in candidateChildTypes)
+            {
+                try
+                {
+                    var deserializedObject = serializer.Deserialize(jsonObject.CreateReader(), candidateChildType.Type);
+                    if (deserializedObject != null)
+                    {
+                        candidateChildType.DeserializedObject = deserializedObject;
```

Comment:
Recommend doing a yield return of a new CandidateChildType object here and IEnumerable<CandidateChildType> overall rather than modifying existing objects. This gives us immutability and makes this thread safe as well. Though that's unlikely to come up here, it's still good practice.